File: application/app/models/task_definition.rb
class TaskDefinition < ApplicationRecord
  include AwsAccessible

  has_many :tasks

  validates :repository, presence: true, uniqueness: true
  validates :name, presence: true

  class << self
    def register!(repository, params)
      ecs = Aws::ECS::Client.new(region: Settings.aws.region)
      response = ecs.register_task_definition(params)
      create!(
        repository: repository,
        name: response.dig(:task_definition, :family),
      )
    end
  end

  def exist?
    ecs.list_task_definition_families.families.include?(name)
  end

  def run(revieee_target)
    task_arn = run_task(revieee_target)
    if task_arn
      task = tasks.build(
        arn: task_arn,
        pr_number: revieee_target.pr_number,
      )
      task if task.save
    end
  end

  private

  def run_task(revieee_target)
    ecs.run_task(
      cluster: Settings.aws.ecs.cluster_name,
      task_definition: latest_arn,
      overrides: {
        container_overrides: [
          {
            name: 'main',
            environment: [
              {
                name: 'BRANCH',
                value: revieee_target.branch,
              },
            ],
          },
        ],
      },
    ).dig(:tasks, 0, :task_arn)
  end

  def latest_arn
    ecs.list_task_definitions(
      family_prefix: name,
      status: 'ACTIVE',
      sort: 'DESC',
      max_results: 1,
    ).dig(:task_definition_arns, 0)
  end
end

File: infrastructure/functions/src/entity/TaskDefinition.ts
import { Entity, Column, PrimaryGeneratedColumn, CreateDateColumn, UpdateDateColumn } from "typeorm";
import { ECS } from "aws-sdk";
import { RevieeeTarget } from "./RevieeeTarget";
import { Task } from "./Task";

@Entity("task_definitions")
export class TaskDefinition {
    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    repository: string;

    @Column()
    name: string;

    @CreateDateColumn({ name: "created_at" })
    createdAt: string;

    @UpdateDateColumn({ name: "updated_at" })
    updatedAt: string;

    async runEcsTask(revieeeTarget: RevieeeTarget): Promise<string> {
        const latestArn = await this.getLatestArn();
        const params: AWS.ECS.RunTaskRequest = {
            cluster: process.env.CLUSTER_NAME,
            taskDefinition: latestArn,
            overrides: {
                containerOverrides: [{
                    name: "main",
                    environment: [{
                        name: "BRANCH",
                        value: revieeeTarget.branch,
                    }],
                }],
            },
        };

        const ecs = new ECS();
        return ecs.runTask(params).promise()
            .then((result: ECS.RunTaskResponse) => {
                return result.tasks[0].taskArn;
            });
    }

    private async getLatestArn(): Promise<string> {
        const params: ECS.ListTaskDefinitionsRequest = {
            familyPrefix: this.name,
            status: "ACTIVE",
            sort: "DESC",
            maxResults: 1,
        };

        const ecs = new ECS();
        return ecs.listTaskDefinitions(params).promise()
            .then((result: ECS.ListTaskDefinitionsResponse) => {
                return result.taskDefinitionArns[0];
            });
    }
}
